alibi_detect/utils/_types.py

```diff
@@ -15,57 +11,18 @@
 else:
     from typing_extensions import Literal  # noqa
 
-# NumPy NDArray pydantic validator type
-# The code below is adapted from the great https://github.com/cheind/pydantic-numpy repo.
-# TODO - above repo also has some very clever functionality to allow passing the numpy array as a npy/npz file.
-#  This could be useful to explore in order to simplify our loading submodule.
-T = TypeVar("T", bound=np.generic)
-if NumpyVersion(np.__version__) < "1.22.0" or sys.version_info < (3, 9):
-    class NDArray(Generic[T], np.ndarray):
-        """
-        A Generic pydantic model to validate (and coerce) np.ndarray's.
-        """
-        @classmethod
-        def __get_validators__(cls):
-            yield cls.validate
-
-        @classmethod
-        def validate(cls, val: Any, field: ModelField) -> np.ndarray:
-            return _validate(cls, val, field)
-
-else:
-    class NDArray(Generic[T], np.ndarray[Any, T]):  # type: ignore[no-redef, type-var]
-        """
-        A Generic pydantic model to validate (and coerce) np.ndarray's.
-        """
-        @classmethod
-        def __get_validators__(cls):
-            yield cls.validate
-
-        @classmethod
-        def validate(cls, val: Any, field: ModelField) -> Optional[np.ndarray]:
-            return _validate(cls, val, field)
-
-
-def _validate(cls: Type, val: Any, field: ModelField) -> np.ndarray:
-    if field.sub_fields is not None:
-        dtype_field = field.sub_fields[0]
-        return np.asarray(val, dtype=dtype_field.type_)
-    else:
-        return np.asarray(val)
-
 
 # Optional dep dependent tuples of types
-supported_models_tf, supported_models_torch, supported_models_sklearn = (), (), ()  # type: ignore
-supported_optimizers_tf, supported_optimizers_torch = (), ()  # type: ignore
+supported_models_tf, supported_models_torch = None, None
+supported_optimizers_tf, supported_optimizers_torch = None, None
 if has_tensorflow:
     import tensorflow as tf
-    supported_models_tf = (tf.keras.Model, )  # type: ignore
-    supported_optimizers_tf = (tf.keras.optimizers.Optimizer, )  # type: ignore
+    supported_models_tf = tf.keras.Model
+    supported_optimizers_tf = (tf.keras.optimizers.Optimizer, type)
 if has_pytorch:
     import torch
-    supported_models_torch = (torch.nn.Module, torch.nn.Sequential)  # type: ignore
```

Comment:
`nn.Sequential` is a subclass of `nn.Module` hence `nn.Sequential` removed.

Reply:
Good point. issue opened: https://github.com/SeldonIO/alibi-detect/issues/668

Fair point about the naming now... I thought it might be worth keeping as plural for consistency with `supported_models_all`, and in case we need to make them collections again in the future. Can change to singular if you think its better though...

Reply:
Why not keep them tuples of one member, e.g. `(torch.nn.Module, )`? Would keep consistency and perhaps extensibility (if it turns out, for any reason, that we might only support a subset of model types.